beamer/health/notify.py

```diff
@@ -0,0 +1,105 @@
+import json
+import os
+import time
+from pathlib import Path
+from typing import Callable, Optional, TypedDict
+
+import requests
+from typing_extensions import NotRequired
+
+
+_TIMEOUT_BETWEEN_SENDS = 61
+
+
+NotificationConfig = dict[str, dict[str, str]]
+
+
+class Message(TypedDict):
+    text: str
+    message_link: NotRequired[str]
+
+
+class NotificationState:
+    _cache_path: Path
+    _notified_state: dict[str, dict[str, bool]]
+
+    def __init__(self, cache_path: Path) -> None:
+        self._cache_path = cache_path
+
+        try:
+            with open(os.path.join(self._cache_path, "breadcrumbs.json"), "r") as f:
+                self._notified_state = json.load(f)
+        except FileNotFoundError:
+            self._notified_state = {}
+
+    def is_set(self, request_id: str, notified_on: str) -> bool:
+        breadcrumb = self._notified_state.get(str(request_id))
+        if breadcrumb is None:
+            return False
+
+        return breadcrumb.get(notified_on, False)
+
+    def update(self, request_id: str, notified_on: str) -> None:
+        breadcrumb = self._notified_state.get(request_id)
+        if breadcrumb is None:
+            self._notified_state[request_id] = {notified_on: True}
+        else:
+            breadcrumb[notified_on] = True
+
+    def persist(self) -> None:
+        with open(os.path.join(self._cache_path, "breadcrumbs.json"), "w") as f:
+            json.dump(self._notified_state, f)
+
+
+class Notify:
+    notifications_sent = 0
+
+    def __init__(self, system: str, config: NotificationConfig) -> None:
+        self.system = system
+        self.config = config
+
+    def send(self, message: Message, callback: Optional[Callable] = None) -> None:
+        if self.notifications_sent > 0 and self.notifications_sent % 5 == 0:
+            time.sleep(_TIMEOUT_BETWEEN_SENDS)
+
+        text = message["text"]
+
+        message_link = message.get("message_link")
+        if message_link is not None:
+            text += f"\nView on [explorer]({message_link})"
+
+        match self.system:
+            case "rocketchat":
+                status = self.send_to_rocketchat(text)
+            case "telegram":
+                status = self.send_to_telegram(text)
+            case _:
+                raise ValueError("Unknown notification system")
+
+        if status:
+            self.notifications_sent += 1
+            if callback is not None:
+                callback()
+
+    def send_to_rocketchat(self, text: str) -> bool:
+        headers = {"Content-Type": "application/json"}
+
+        body = {
+            "username": "beamer-monitoring-bot",
```

Comment:
A really small nitpick: maybe update the `monitoring` part since we moved away from that wording...